### PR TITLE
feat: guard function for traceful + option

### PR DESCRIPTION
**Pull request by cwaldm**
I wanted to have a `guard` equivalent for `traceful (option a)` functions to simplify protocol functions.
@qaphla found the solution of adding `b` to the post-condition of the option `guard`.
This is technically the only change we need, but for convenience I also added a `guard_tr` function hiding the `return`.

With this, we can now write 
``` fstar
let send_msg1 global_sess_id alice sess_id =
  let*? st: nsl_session = get_state alice sess_id in
  guard_tr (InitiatorSentMsg1? st);*?
  let InitiatorSentMsg1 bob n_a = st in
  ...
```
in the NSL example, as opposed to having to write a match and deal with the None cases:
``` fstar
let send_msg1 global_sess_id alice sess_id =
  let*? st: nsl_session = get_state alice sess_id in
  match st with
  | InitiatorSentMsg1 bob n_a -> (
       ... 
  )
  | _ -> return None
```

(I only used the new guard function in `send_msg1` of NSL to illustrate its use. If you want, I can also change all of the example protocol functions to use it.)

### CODE REVIEW

File: src/core/DY.Core.Trace.Manipulation.fst [hunk 1]
```diff
@@ -73,11 +73,19 @@ let get_trace tr =
 
 /// guard function for the option monad.
 
-val guard: b:bool -> option unit
+val guard: b:bool -> option (u:unit{b})
```
Comment (TWal) [inline src/core/DY.Core.Trace.Manipulation.fst:76]:
```suggestion
val guard: b:bool -> option (squash b)
```
nit: there is already a standard name for `u:unit{b}`

File: src/core/DY.Core.Trace.Manipulation.fst [hunk 2]
```diff
@@ -73,11 +73,19 @@ let get_trace tr =
...
 let guard b =
-  if b then Some ()
+  if b
+  then Some ()
   else None
 
+/// guard function for the traceful + option monad
+
+val guard_tr : (b:bool) -> traceful (option (u:unit{b}))
```
Comment (TWal) [inline src/core/DY.Core.Trace.Manipulation.fst:84]:
```suggestion
val guard_tr : (b:bool) -> traceful (option (squash b))
```
same nit as above

**Comment by TWal** [comment]
This is definitely a useful change, thanks!
Before merging it would be nice to convert the rest of the example to use this style, for consistency.

About the git thing, I am wondering why we see commit 667b296ba7e4e51c619bd49a5a3f76bfbb83de9a in the commits associated with this PR?

**Comment by cwaldm** [comment]
Ok. I'll update the NSL example to use the new guard everywhere.
Do you also want the same for the iso-dh example?

**Comment by TWal** [comment]
If you have the energy to do so, it would be better yes.

About the git commit thing, I guess no action is needed because `main` is merged and git shows no merge-conflict to merge this PR, so it should be good.

**Comment by cwaldm** [comment]
I updated both NSL and ISO-DH (Protocol.Stateful.fst) to use `guard_tr` instead of the matches

**Review by TWal — APPROVED** [review]
Looks good, thanks!

**Pull request merged**